tests/Integration/Behaviour/Features/Context/Domain/Product/ProductImageFeatureContext.php

```diff
@@ -398,13 +404,77 @@ private function uploadImageByShopConstraint(string $imageReference, string $fil
     /**
      * @param string $productReference
      * @param TableNode $tableNode
-     * @param array $expectedShopIds
      *
      * @return void
      */
-    private function assertProductImagesForShops(string $productReference, TableNode $tableNode, array $expectedShopIds)
+    private function assertAllProductImages(string $productReference, TableNode $tableNode)
     {
         $images = $this->getProductImages($productReference);
+        $this->assertProductImages($tableNode, $images);
+    }
+
+    /**
+     * @param string $productReference
+     * @param TableNode $tableNode
+     * @param int $shopId
+     *
+     * @return void
+     */
+    private function assertProductImagesByShopId(string $productReference, TableNode $tableNode, int $shopId)
+    {
+        $images = array_filter(
+            $this->getProductImages($productReference),
+            static function (ProductImage $productImage) use ($shopId): bool {
+                return in_array($shopId, $productImage->getShopIds(), true);
+            }
+        );
+        $this->assertProductImages($tableNode, $images);
+    }
+
+    /**
+     * @When /^I apply the following matrix of images for product "([^"]*)":$/
+     */
+    public function iApplyTheFollowingMatrixOfImagesForProduct(string $productReference, TableNode $table)
+    {
+        $command = new SetProductImagesForAllShopCommand(
+            $this->getSharedStorage()->get(trim($productReference))
+        );
+        foreach ($table as $data) {
```

Comment:
oh I didn't know table is iterable. I thought we first need to call ->getRowsHash or ->getColumnsHash :smile: But I see it now.. ok :+1: